Add Efficient Attention support on ROCM

xinyazhang opened the pull request:
This patch implements `with sdpa_kernel(SDPBackend.EFFICIENT_ATTENTION):` by reusing AOTriton's accelerated SDPA implementation

Known limitations:
- Only supports MI200/MI300X GPUs
- Does not support varlen
- Does not support `CausalVariant`
- Optional arguments `causal_diagonal` and `seqlen_k` in `_efficient_attention_forward/backward` must be null
- Does not work well with inductor's SDPA rewriter. The rewriter has been updated to only use math and flash attention on ROCM.

This PR also uses a different approach of installing AOTriton binary instead of building it from source in the base docker image. More details on motivation: https://github.com/pytorch/pytorch/pull/124885#issuecomment-2153229129

`PYTORCH_TEST_WITH_ROCM=1 PYTORCH_TESTING_DEVICE_ONLY_FOR="cuda" python test/test_transformers.py` yields "55028 passed, 20784 skipped" results with this change.  [Previous result](https://hud.pytorch.org/pr/127528) of `test_transformers.py` was 0 error, 0 failure, 55229 skipped out of 75517 tests in total (the XML report does not contain total number of passed tests).

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

xinyazhang committed:
move mk_aotensor to aotriton_adapter.h

xinyazhang committed:
MI300X arch typo fix

xinyazhang committed:
Use aotriton to handle efficient attention requests.

xinyazhang committed:
enable meff on supported ROCM systems.

xinyazhang committed:
Update can_use_mem_efficient_attention for ROCM

xinyazhang committed:
Add bias gradients support

xinyazhang committed:
properly unselect float32

xinyazhang committed:
Implement _fill_mem_eff_dropout_mask_ for AOTriton

xinyazhang committed:
test_transformers.py: enable dropout and attn_mask gradients tests for meff

xinyazhang committed:
Fix typo

xinyazhang committed:
Fix grad_bias.

This tensor does not need transpose.

xinyazhang committed:
Skip failed UTs

xinyazhang committed:
Skip torch.use_deterministic_algorithms

xinyazhang committed:
disable fp32 SDPA ops on ROCM

xinyazhang committed:
Enable more SDPA ops on ROCM

xinyazhang committed:
flash_api.hip: fix empty_bias's dtype.

This is not strictly required for current AOTriton, but may be necessary
in future versions.

xinyazhang commented:
@pytorchbot label "ciflow/rocm"

xinyazhang committed:
Fix missing cast_dtype

xinyazhang committed:
Fix unused variable on CUDA platform.

xinyazhang committed:
Revert "disable fp32 SDPA ops on ROCM"

This reverts commit 11e18c275334d82bf12dc2a44eaea138ddeaa377.

xinyazhang committed:
Enable fp32 again since newer AOTriton fixed it.

xinyazhang committed:
Add fp32 support fp32 for efficient attention

Also includes some tweaks in test_transformers.py

xinyazhang committed:
unskip test_mem_eff_attention_non_contiguous_mask

xinyazhang committed:
Skip failed UTs for efficient attention

xinyazhang committed:
Fix lint

xinyazhang commented:
@pytorchbot label "keep-going"

xinyazhang committed:
remove the debugging static_assert

xinyazhang committed:
Disable failed tests

xinyazhang committed:
Use latest AOTriton main (refactor the build system)

xinyazhang committed:
Fix distributed/_tensor/test_attention.py::test_ring_attention_compile_attention

1. The Rank of logsumexp Tensor must be 3. This tensor was considered
for internal use only but apparently exposed to UTs.

2. Also includes a fix for mGPU. The stream should be selected after
picking the current device according to input tensor.

3. The UT does not skip efficient attention operator on platforms that do
not support efficent attention. Support of flash attention does not
always ensure efficient attention is implemented.

xinyazhang committed:
Skip test_ring_attention_sdpa

_c10d_functional_autograd::all_to_all_single seems not implemented on ROCm.

Note: another unfixed problem is the mismatching of outputs between
torch.ops.aten._scaled_dot_product_flash_attention and
_scaled_dot_product_chunk_flash_attention. We need fix both problems to
enable this UT.

xinyazhang committed:
Fix lint

jithunnair-amd commented:
@malfet Please review. All ROCm CI jobs passed.

malfet reviewed (approved):
Check that it only affects ROCM parts without checking what it actually changes in the hip files

malfet commented:
@pytorchbot merge

xinyazhang committed:
Restore the behavior of RingAttentionTest.test_ring_attention_compile on CUDA platform.

On CUDA platform, this test bundles memory efficient attention (ME) test
with flash attention (FA) test, and on platforms that do not suppport
FA, it will not test ME as well.

On ROCM we hope they can be tested independently, and refactored the
skip conditions, but this apparently enabled ME tests accidentally and
caused undiscovered failures on CUDA platforms.

This patch restores the old testing behavior on CUDA (non-ROCM) platform.

xinyazhang commented:
I think I know why the UT failed on CUDA. See the commit message of https://github.com/pytorch/pytorch/pull/124885/commits/0f8257af71d8763b4621e41894b79617f0c8e112

pruthvistony commented:
@pytorchbot rebase

xinyazhang commented:
@pruthvistony do NOT rebase against viable/strict. It does not contain necessary commit that moves AOTriton to base docker image.

malfet commented:
@pytorchbot merge -i "Claim is that it should not introduce new regressions"

jithunnair-amd commented:
@malfet Thanks for the approval! We are waiting for a green signal on the ROCm CI jobs (they timed out in the last attempt) before merging. We'll handle the merge once signal is green.

xinyazhang commented:
> @xinyazhang The only ROCm CI job currently failing is https://github.com/pytorch/pytorch/actions/runs/9379259480/job/25848946879

`grep FAILED job-logs.txt` shows
```
2024-06-05T16:22:26.0905686Z inductor/test_compiled_optimizers.py::CompiledOptimizerParityTestsCUDA::test_correctness_Adadelta_use_closure_True_cuda_float32 FAILED [14.9466s] [ 98%]
2024-06-05T16:22:26.2341935Z FAILED [14.9466s] inductor/test_compiled_optimizers.py::CompiledOptimizerParityTestsCUDA::test_correctness_Adadelta_use_closure_True_cuda_float32 - torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
2024-06-05T16:34:56.7964726Z inductor/test_efficient_conv_bn_eval.py::EfficientConvBNEvalCudaTests::test_basic_cuda FAILED [47.7058s] [100%]
2024-06-05T16:34:56.8509340Z FAILED [47.7058s] inductor/test_efficient_conv_bn_eval.py::EfficientConvBNEvalCudaTests::test_basic_cuda - torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
2024-06-05T16:34:56.8528608Z inductor/test_efficient_conv_bn_eval.py::EfficientConvBNEvalCudaTests::test_basic_cuda FAILED [84.0695s] [100%]
2024-06-05T16:34:56.8889705Z FAILED [84.0695s] inductor/test_efficient_conv_bn_eval.py::EfficientConvBNEvalCudaTests::test_basic_cuda - torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
2024-06-05T16:34:56.8908006Z inductor/test_efficient_conv_bn_eval.py::EfficientConvBNEvalCudaTests::test_basic_cuda FAILED [0.9633s] [100%]
2024-06-05T16:34:56.9167623Z FAILED [0.9633s] inductor/test_efficient_conv_bn_eval.py::EfficientConvBNEvalCudaTests::test_basic_cuda - torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
```

Consider we have the `keep-going` label. This is the only failure, and it seems irrelevant to this PR.
(We may want to disable this UT, tho)

xinyazhang committed:
Install AOTriton through GitHub release tarball file.

xinyazhang committed:
Fix install_aotriton.sh and related Dockerfile

xinyazhang committed:
Move the install of AOTriton to later stage.

jithunnair-amd commented:
Some observations regarding AOTriton build:
* Originally, this PR was continuing with the approach of building AOTriton from source in the base docker image 
* The build time for AOTriton on a linux.12x.large machine increased from [20 min](https://github.com/pytorch/pytorch/pull/127012#issue-2313836750) to [30 min](https://github.com/pytorch/pytorch/actions/runs/9379257816/job/25823929540)
* However, we found that in case PyTorch builds jobs are triggered in parallel with the docker builds jobs, the former will try to build the base docker image from scratch if the latter hasn't pushed the docker image yet. This means that the PyTorch build job will try to build aotriton on a linux.2x.large machine and time out at [4hrs](https://github.com/pytorch/pytorch/actions/runs/9379259480/job/25823912105)
* So we now install aotriton from a release tarball in the base docker image: `https://github.com/ROCm/aotriton/releases/download/${VER}/aotriton-${VER}-${MANYLINUX}_${ARCH}-${ROCMBASE}.tar.bz2`
* This reduces the base docker image build time from [1hr 44min](https://github.com/pytorch/pytorch/actions/runs/9379257816/job/25823929540) to [58 min](https://github.com/pytorch/pytorch/actions/runs/9404529885/job/25903670744)
* This also reduces the PyTorch build job duration (when building the base docker image from scratch) to [2hrs](https://github.com/pytorch/pytorch/actions/runs/9404531458/job/25903651472) (1hr 40 mins for building the base docker image on a linux.2x.large node)

jithunnair-amd commented:
Adding @huydhn to review the changes wrt AOTriton build.
Adding @atalman as well since we plan to make these same changes in the wheel build flow as well by installing AOTriton from binary.

drisspg commented on aten/src/ATen/native/transformers/cuda/attention.cu:
```diff
@@ -1063,6 +1070,64 @@ std::tuple<Tensor, Tensor, Tensor, Tensor, c10::SymInt, c10::SymInt> _efficient_
 
+#ifdef USE_ROCM
+  // ROCM Implementation
+  auto ret = aotriton::v2::flash::check_gpu(stream);
```
do you need this if the check is already in sdp_utils?

xinyazhang replied:
Oh I'm following the practice of `mha_fwd`, which also perform the checks for GPU architecture.
https://github.com/pytorch/pytorch/blob/dc8e3c2e904f22f05646e1d87987459af8845d7b/aten/src/ATen/native/transformers/cuda/flash_attn/flash_api.cpp#L359-L363

I'm not so sure what will happen if SDPA is call on GPUs with different arches (e.g., MI200+Navi31). This check is the last line of defense to guarantee the target GPU does support the SDPA request.

drisspg commented on test/distributed/_tensor/test_attention.py:
```diff
@@ -299,18 +307,29 @@ def test_ring_attention_custom_transformer(self) -> None:
         ],
     )
     def test_ring_attention_compile(self, attention_fn: object) -> None:
+        if attention_fn is None:
```
is this another, "parametrize cant be empty" problem again?

xinyazhang replied:
Passing `None` as `attention_fn` is a trick to ensure on every platform, `test_ring_attention_compile_attention_fn0` always refers to the test of FA, and `test_ring_attention_compile_attention_fn0` is always for ME.

(However I do agree the skipping message can be improved, but haven't found an elegant way)

drisspg commented on test/inductor/test_fused_attention.py:
```diff
@@ -315,6 +317,7 @@ def dot_prod_attention(
             checkpoint_wrapper(dot_prod_attention), contains=False, has_dropout=True
         )
 
+    @skipIfRocm  # AssertionError: expected size 4==4, stride 32==64 at dim=0
```
Is this going to cause issues for people using compile + rocm?

xinyazhang replied:
We fixed this in commit https://github.com/pytorch/pytorch/pull/124885/commits/0ff88e2109ec2a4a2f2b5ced54c84b3a2dee9596, by forcing the rewriter to use `aten._scaled_dot_product_attention_math` on ROCM

drisspg commented on test/test_transformers.py:
```diff
@@ -2572,13 +2576,13 @@ def test_mem_eff_backwards_determinism(self, device):
-    @parametrize("seq_len_q", [4, 8, 64, 128, 256, 512, 1024, 2048] if SM80OrLater else [4, 8, 64, 128, 256, 512])
-    @parametrize("seq_len_k", [4, 8, 64, 128, 256, 512, 1024, 2048] if SM80OrLater else [4, 8, 64, 128, 256, 512])
-    @parametrize("head_dim", [8, 16, 32, 64, 72, 96, 128] if SM80OrLater else [8, 16, 32, 64])
+    @parametrize("seq_len_q", [4, 8, 64, 128, 256, 512, 1024, 2048] if SM80OrLater or TEST_WITH_ROCM else [4, 8, 64, 128, 256, 512])
```
maybe we should extract this to something like, `platform supports long sequences`?

xinyazhang replied:
Replaced with `MEM_EFF_CAPABILITY_MATCHES_SM80` in the latest commit

drisspg commented on test/test_transformers.py:
```diff
@@ -2591,6 +2595,8 @@ def _get_mem_eff_drop_mask(batch_size, n_heads, q_len, kv_len, p, seed, offset,
         if max(seq_len_q, seq_len_k) >= 2048 and torch.cuda.get_device_properties('cuda').total_memory < 40 * 2**30:
             unittest.skip("Reference implementation OOM")
             return
+        if TEST_WITH_ROCM and seq_len_q * seq_len_k * head_dim * batch_size > 1024 * 1024 * 128:
```
hmm but this actually is a counter point to the above?

drisspg commented on test/test_transformers.py:
```diff
@@ -2694,6 +2702,11 @@ def _get_mem_eff_drop_mask(batch_size, n_heads, q_len, kv_len, p, seed, offset,
+            unittest.skip("Skip fp32 attn_mask gradients on ROCM, for now.")
+            return
+        if TEST_WITH_ROCM and seq_len_q * seq_len_k * head_dim * batch_size > 1024 * 1024 * 128:
+            torch.cuda.empty_cache()  # Prevent memory fragmentation
```
out of curiosity what happens if you dont call this? will this oom?

xinyazhang replied:
Yes. This will reliably cause OOM on MI210 systems when testing with the whole suite (but cannot be reproduced with any individual test).

drisspg reviewed (commented):
left some overarching comments, nothing looks blocking though

xinyazhang committed:
Replace SM80OrLater or TEST_WITH_ROCM with MEM_EFF_CAPABILITY_MATCHES_SM80

xinyazhang committed:
Lower SDPA patterns to aten._scaled_dot_product_attention_math on ROCM.

aten.scaled_dot_product_attention uses efficient attention if available,
but the backend does not work with inductor for now.

This allows us to enable test_sdpa_rewriter_5_cuda and test_sdpa_rewriter_12_cuda.
However, test_sdpa_rewriter_3_cuda and test_sdpa_rewriter_4_cuda are
still disabled (See https://github.com/pytorch/pytorch/issues/122896)

xinyazhang committed:
Fix python test/dynamo/test_activation_checkpointing.py -k ActivationCheckpointingViaTagsTests.test_pattern_matcher

Need to enable FA but disable ME.

xinyazhang committed:
Fix lint

jithunnair-amd commented:
@pytorchbot merge
